[WIP] Dropdown bug fixes

rebeccaalpert opened the pull request:
Related to [CONSOLE-519](https://jira.coreos.com/browse/CONSOLE-519).

Namespace dropdown now works with tabbing and <a> elements have hrefs. Filtered list now works correctly.

spadgett commented on frontend/public/components/utils/dropdown.jsx:
```diff
@@ -114,25 +114,33 @@ class DropDownRow extends React.PureComponent {
     let prefix;
     if (onUnBookmark) {
-      prefix = <a className={classNames('bookmarker', {focus: selected, hover})} onClick={e => onUnBookmark(e, itemKey)}><i className="fa fa-minus-circle" /></a>;
+      prefix = <a href={url} className={classNames('bookmarker', {focus: selected, hover})} onClick={e => onUnBookmark(e, itemKey)}><i className="fa fa-minus-circle" /></a>;
```
We don't want to make them real links. We should use something like `href="#"` and then call `e.preventDefault()`. This makes it keyboard focusable without changing the behavior.

spadgett replied:
Another fix that might work is giving `tabIndex=0` since we're already handling keyboard events

spadgett commented:
/assign

rebeccaalpert commented:
I adjusted the hrefs and CSS mentioned here: https://jira.coreos.com/projects/CONSOLE/issues/CONSOLE-701

I also fixed up the focus situation.

spadgett commented on frontend/public/components/utils/dropdown.jsx:
```diff
@@ -118,21 +118,21 @@ class DropDownRow extends React.PureComponent {
     let prefix;
     if (onUnBookmark) {
-      prefix = <a className={classNames('bookmarker', {focus: selected, hover})} onClick={e => onUnBookmark(e, itemKey)}><i className="fa fa-minus-circle" /></a>;
+      prefix = <a className={classNames('bookmarker', {hover, focus: selected})} onClick={e => onUnBookmark(e, itemKey)}><i className="fa fa-minus-circle" /></a>;
```
These links should have an `href` also. The icons should have `aria-hidden="true"`

spadgett commented on frontend/public/components/utils/dropdown.jsx:
```diff
@@ -118,21 +118,21 @@ class DropDownRow extends React.PureComponent {
+    return <li role="option" tabIndex={(selected || hover) ? 0 : null} className={classNames(className)} key={itemKey}>
       {prefix}
-      <a ref={this.link} id={`${itemKey}-link`} className={classNames({'next-to-bookmark': !!prefix, focus: selected, hover})} onClick={e => onclick(itemKey, e)}>{content}</a>
+      <a href="#" ref={this.link} id={`${itemKey}-link`} className={classNames({'next-to-bookmark': !!prefix, hover, focus: selected})} onClick={e => onclick(itemKey, e)}>{content}</a>
```
Just confirming we're calling `e.preventDefault()` in the on click handler

rebeccaalpert commented:
We were already calling e.preventDefault() in the general click handler; I added it in for the favorite and bookmark click functions. Addressed other feedback too.

spadgett commented on frontend/public/components/namespace.jsx:
```diff
@@ -270,6 +270,7 @@ class NamespaceDropdown_ extends React.Component {
         className="co-namespace-selector__dropdown"
         menuClassName="co-namespace-selector__menu"
         noButton
+        namespace
```
@rebeccaalpert I'm still not clear why we're special casing the namespace dropdown. The dropdown component shouldn't need to know anything about the namespace picker.

rebeccaalpert commented:
Removed.

spadgett commented:
When you have bookmarks and an active text filter, the behavior isn't quite right. It doesn't treat the bookmark as being at the top of the list.

rebeccaalpert commented:
Yikes. It looks okay with this change.

spadgett commented:
It's a little confusing that there are two highlighted items if you use the keyboard then also hover over an item with the mouse. I think it's OK to open a follow on issue to address, though.

![screen shot 2018-08-17 at 2 16 40 pm](https://user-images.githubusercontent.com/1167259/44282211-526bfb00-a228-11e8-86d7-9c22edeb2bb5.png)

rebeccaalpert commented:
Addressed feedback.

spadgett commented on frontend/public/components/utils/dropdown.jsx:
```diff
@@ -287,6 +291,21 @@ export class Dropdown extends DropdownMixin {
+    }
+
+    // so bookmarks advance correctly
+    if (bookmarks) {
```
This is probably OK as-is, but more robust if we use `_.isEmpty`

rebeccaalpert commented:
Squashed.

spadgett commented:
@sg00dwin FYI, not sure if this conflicts with your PR.

rebeccaalpert commented:
jenkins rebuild

spadgett commented:
jenkins rebuild
